File: server/src/main/java/org/e8yes/service/UserService.java
/**
 * e8yes demo web server.
 *
 * <p>Copyright (C) 2020 Chifeng Wen {daviesx66@gmail.com}
 *
 * <p>This program is free software: you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * <p>You should have received a copy of the GNU General Public License along with this program. If
 * not, see <http://www.gnu.org/licenses/>.
 */
package org.e8yes.service;

import com.google.protobuf.ByteString;
import io.grpc.Status;
import io.grpc.stub.StreamObserver;
import java.lang.reflect.InvocationTargetException;
import java.sql.SQLException;
import java.util.List;
import java.util.Optional;
import java.util.logging.Level;
import java.util.logging.Logger;
import org.e8yes.environment.Initializer;
import org.e8yes.exception.AccessDeniedException;
import org.e8yes.exception.ResourceMissingException;
import org.e8yes.service.identity.JwtAuthorizer;
import org.e8yes.service.identity.UserCreation;
import org.e8yes.service.identity.UserEntity;
import org.e8yes.service.identity.UserProfile;
import org.e8yes.service.identity.UserRetrieval;
import org.e8yes.util.PaginationValidator;

/** Service for user management. */
public class UserService extends UserServiceGrpc.UserServiceImplBase {

  @Override
  public void register(RegistrationRequest req, StreamObserver<RegistrationReponse> res) {
    RegistrationReponse response;
    UserEntity user;
    try {
      user =
          UserCreation.createBaselineUser(
              req.getSecurityKey().toByteArray(),
              /*userId=*/ Optional.empty(),
              Initializer.environmentContext().demowebDbConnections().connectionReservoir());
      response =
          RegistrationReponse.newBuilder()
              .setErrorType(RegistrationReponse.RegistrationErrorType.RET_NoError)
              .setUserId(user.id.value())
              .build();

      res.onNext(response);
      res.onCompleted();
    } catch (SQLException | IllegalAccessException ex) {
      Logger.getLogger(UserService.class.getName()).log(Level.SEVERE, null, ex);
      res.onError(Status.INTERNAL.withDescription(ex.getMessage()).asException());
    }
  }

  @Override
  public void authorize(AuthorizationRequest req, StreamObserver<AuthorizationResponse> res) {
    try {
      UserEntity user =
          UserRetrieval.retrieveUserEntity(
              req.getUserId(),
              Initializer.environmentContext().demowebDbConnections().connectionReservoir());
      if (user == null) {
        throw new ResourceMissingException("User with ID=" + req.getUserId() + " doesn't exist.");
      }
      JwtAuthorizer.AuthorizationToken token =
          JwtAuthorizer.signAuthorizationToken(
              user,
              req.getSecurityKey().toByteArray(),
              Initializer.environmentContext().authorizationJwtProvider().algorithm());

      res.onNext(
          AuthorizationResponse.newBuilder()
              .setJwtToken(ByteString.copyFrom(token.jwtToken))
              .build());
      res.onCompleted();
    } catch (SQLException
        | NoSuchMethodException
        | InstantiationException
        | IllegalAccessException
        | IllegalArgumentException
        | InvocationTargetException ex) {
      Logger.getLogger(UserService.class.getName()).log(Level.SEVERE, null, ex);
      res.onError(Status.INTERNAL.withDescription(ex.getMessage()).asException());
    } catch (ResourceMissingException ex) {
      res.onError(Status.NOT_FOUND.withDescription(ex.getMessage()).asException());
    } catch (AccessDeniedException ex) {
      res.onError(Status.UNAUTHENTICATED.withDescription(ex.getMessage()).asException());
    }
  }

  @Override
  public void getPublicProfile(
      GetPublicProfileRequest req, StreamObserver<GetPublicProfileResponse> res) {}

  @Override
  public void updatePublicProfile(
      UpdatePublicProfileRequest req, StreamObserver<UpdatePublicProfileResponse> res) {}

  @Override
  public void search(SearchUserRequest req, StreamObserver<SearchUserResponse> res) {
    try {
      PaginationValidator.validate(req.getPagination(), /*resultPerPageLimit=*/ 100);
      Optional<Long> userIdPrefix =
          req.getUserId() == null || req.getUserId().getValue() == 0
              ? Optional.empty()
              : Optional.of(req.getUserId().getValue());
      Optional<String> aliasPrefix =
          req.getAlias() == null || req.getAlias().getValue().isEmpty()
              ? Optional.empty()
              : Optional.of(req.getAlias().getValue());

      List<UserEntity> results =
          UserRetrieval.searchUserEntity(
              userIdPrefix,
              aliasPrefix,
              req.getPagination(),
              Initializer.environmentContext().demowebDbConnections().connectionReservoir());

      SearchUserResponse.Builder builder = SearchUserResponse.newBuilder();
      for (UserEntity user : results) {
        UserPublicProfile profile = UserProfile.extractPublicInfo(user);
        builder.addEntries(profile);
      }

      res.onNext(builder.build());
      res.onCompleted();
    } catch (SQLException
        | NoSuchMethodException
        | InstantiationException
        | IllegalAccessException
        | InvocationTargetException ex) {
      Logger.getLogger(UserService.class.getName()).log(Level.SEVERE, null, ex);
      res.onError(Status.INTERNAL.withDescription(ex.getMessage()).asException());
    } catch (IllegalArgumentException ex) {
      res.onError(Status.INVALID_ARGUMENT.withDescription(ex.getMessage()).asException());
    }
  }
}

File: demoweb_service/demoweb/service/user_service.cc
/**
 * e8yes demo web.
 *
 * <p>Copyright (C) 2020 Chifeng Wen {daviesx66@gmail.com}
 *
 * <p>This program is free software: you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * <p>You should have received a copy of the GNU General Public License along with this program. If
 * not, see <http://www.gnu.org/licenses/>.
 */

#include <cassert>
#include <grpcpp/grpcpp.h>
#include <optional>
#include <string>
#include <unordered_set>
#include <vector>

#include "demoweb_service/demoweb/common_entity/user_entity.h"
#include "demoweb_service/demoweb/constant/pagination.h"
#include "demoweb_service/demoweb/environment/environment_context_interface.h"
#include "demoweb_service/demoweb/module/create_user.h"
#include "demoweb_service/demoweb/module/user_storage.h"
#include "demoweb_service/demoweb/module/search_user.h"
#include "demoweb_service/demoweb/module/user_identity.h"
#include "demoweb_service/demoweb/module/user_profile.h"
#include "demoweb_service/demoweb/service/service_util.h"
#include "demoweb_service/demoweb/service/user_service.h"
#include "proto_cc/pagination.pb.h"
#include "proto_cc/service_user.grpc.pb.h"
#include "proto_cc/service_user.pb.h"
#include "proto_cc/user_relation.pb.h"

namespace e8 {

grpc::Status UserServiceImpl::Register(grpc::ServerContext * /*context*/,
                                       RegistrationRequest const *request,
                                       RegistrationResponse *response) {

    std::optional<UserEntity> user =
        CreateBaselineUser(request->security_key(),
                           /*userId=*/std::nullopt, DemoWebEnvironment()->CurrentHostId(),
                           DemoWebEnvironment()->DemowebDatabase());
    if (!user.has_value()) {
        return grpc::Status(grpc::StatusCode::INTERNAL,
                            "User ID conflicts when it shouldn't happen");
    }

    response->set_user_id(*user->id.Value());
    response->set_error_type(RegistrationResponse::RET_NoError);

    return grpc::Status::OK;
}

grpc::Status UserServiceImpl::Authorize(grpc::ServerContext * /*context*/,
                                        AuthorizationRequest const *request,
                                        AuthorizationResponse *response) {
    std::optional<UserEntity> user =
        FetchUser(request->user_id(), DemoWebEnvironment()->DemowebDatabase());
    if (!user.has_value()) {
        return grpc::Status(grpc::StatusCode::NOT_FOUND,
                            "User ID=" + std::to_string(request->user_id()) + " doesn't exist.");
    }

    std::optional<SignedIdentity> signed_identity =
        SignIdentity(user.value(), request->security_key(), DemoWebEnvironment()->KeyGen());
    if (!signed_identity.has_value()) {
        return grpc::Status(grpc::StatusCode::UNAUTHENTICATED,
                            "Failed to validate the provided security key.");
    }

    IdentitySignature sig;
    sig.set_signature(signed_identity.value());
    *response->mutable_signed_identity() = sig;

    return grpc::Status::OK;
}

grpc::Status UserServiceImpl::GetPublicProfile(grpc::ServerContext *context,
                                               GetPublicProfileRequest const *request,
                                               GetPublicProfileResponse *response) {
    std::optional<Identity> identity = ExtractIdentityFromContext(*context, nullptr);

    std::optional<UserEntity> user =
        FetchUser(request->user_id(), DemoWebEnvironment()->DemowebDatabase());
    if (!user.has_value()) {
        return grpc::Status(grpc::StatusCode::NOT_FOUND,
                            "User ID=" + std::to_string(request->user_id()) + " doesn't exist.");
    }

    std::vector<UserPublicProfile> profiles;
    if (identity.has_value()) {
        profiles = BuildPublicProfiles(identity.value().user_id(), {user.value()},
                                       DemoWebEnvironment()->KeyGen(),
                                       DemoWebEnvironment()->DemowebDatabase());
    } else {
        profiles = BuildPublicProfiles(std::optional<UserId>(), {user.value()},
                                       DemoWebEnvironment()->KeyGen(),
                                       DemoWebEnvironment()->DemowebDatabase());
    }
    assert(profiles.size() == 1);
    *response->mutable_profile() = profiles[0];

    return grpc::Status::OK;
}

grpc::Status UserServiceImpl::UpdatePublicProfile(grpc::ServerContext *context,
                                                  UpdatePublicProfileRequest const *request,
                                                  UpdatePublicProfileResponse *response) {
    grpc::Status status;
    std::optional<Identity> identity = ExtractIdentityFromContext(*context, &status);
    if (!status.ok()) {
        return status;
    }
    UserId user_id = identity.value().user_id();

    std::optional<UserEntity> user = FetchUser(user_id, DemoWebEnvironment()->DemowebDatabase());
    if (!user.has_value()) {
        return grpc::Status(grpc::StatusCode::NOT_FOUND,
                            "User ID=" + std::to_string(user_id) + " doesn't exist.");
    }

    std::optional<std::string> alias =
        request->has_alias() ? std::optional<std::string>(request->alias().value()) : std::nullopt;
    std::optional<std::string> biography =
        request->has_biography() ? std::optional<std::string>(request->biography().value())
                                 : std::nullopt;
    UpdateProfile(alias, biography, &user.value(), DemoWebEnvironment()->DemowebDatabase());

    std::vector<UserPublicProfile> profiles =
        BuildPublicProfiles(user_id, {user.value()}, DemoWebEnvironment()->KeyGen(),
                            DemoWebEnvironment()->DemowebDatabase());
    assert(profiles.size() == 1);
    *response->mutable_profile() = profiles[0];

    return grpc::Status::OK;
}

grpc::Status UserServiceImpl::Search(grpc::ServerContext *context, SearchUserRequest const *request,
                                     SearchUserResponse *response) {
    grpc::Status status = ValidatePagination(request->pagination(), kResultPerPageLimit);
    if (!status.ok()) {
        return status;
    }

    std::optional<Identity> identity = ExtractIdentityFromContext(*context, &status);
    if (!status.ok()) {
        return status;
    }

    std::vector<UserEntity> user_entities =
        SearchUser(identity->user_id(), request->query(),
                   /*oneof_user_relations=*/std::unordered_set<UserRelation>(),
                   request->pagination(), DemoWebEnvironment()->DemowebDatabase());

    std::vector<UserPublicProfile> profiles =
        BuildPublicProfiles(identity->user_id(), user_entities, DemoWebEnvironment()->KeyGen(),
                            DemoWebEnvironment()->DemowebDatabase());

    *response->mutable_user_profiles() = {profiles.begin(), profiles.end()};
    return grpc::Status::OK;
}

grpc::Status UserServiceImpl::PrepareNewAvatar(grpc::ServerContext *context,
                                               PrepareNewAvatarRequest const *request,
                                               PrepareNewAvatarResponse *response) {
    grpc::Status status;
    std::optional<Identity> identity = ExtractIdentityFromContext(*context, &status);
    if (!status.ok()) {
        return status;
    }

    std::optional<UserEntity> user =
        FetchUser(identity.value().user_id(), DemoWebEnvironment()->DemowebDatabase());
    assert(user.has_value());

    if (!AcceptableProfileAvatarFileFormat(request->file_format())) {
        return grpc::Status(grpc::INVALID_ARGUMENT, "Cannot accept the provided file format.");
    }

    AvatarSetup setup =
        SetUpNewProfileAvatar(user.value(), request->file_format(), DemoWebEnvironment()->KeyGen(),
                              DemoWebEnvironment()->DemowebDatabase());
    response->mutable_avatar_readwrite_access()->set_access_token(setup.avatar_path_access_token);

    return grpc::Status::OK;
}

} // namespace e8
